source/extensions/filters/http/jwt_authn/verifier.cc

```diff
@@ -0,0 +1,204 @@
+#include "extensions/filters/http/jwt_authn/verifier.h"
+
+using ::envoy::config::filter::http::jwt_authn::v2alpha::JwtProvider;
+using ::envoy::config::filter::http::jwt_authn::v2alpha::JwtRequirement;
+using ::envoy::config::filter::http::jwt_authn::v2alpha::JwtRequirementAndList;
+using ::envoy::config::filter::http::jwt_authn::v2alpha::JwtRequirementOrList;
+using ::google::jwt_verify::Status;
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace JwtAuthn {
+namespace {
+
+// base verifier for provider_name, provider_and_audiences, and allow_missing_or_failed.
+class BaseVerifierImpl : public Verifier {
+public:
+  void registerParent(Verifier* parent) override { parent_ = parent; }
+
+  void onCompleteHelper(Status status, VerifyContext& context) {
+    if (parent_ != nullptr) {
+      return parent_->onComplete(status, context);
+    }
+    return context.callback()->onComplete(status);
+  }
+
+  virtual void onComplete(const Status& status, VerifyContext& context) override {
+    if (!context.hasResponded(this)) {
+      context.setResponded(this);
+      onCompleteHelper(status, context);
+    }
+  }
+
+protected:
+  // The parent group verifier.
+  Verifier* parent_{};
```

Comment:
This parent_() should come from constructor, it should not change.  

const Verificer* parent_{};

Reply:
done.